numpy/core/fromnumeric.py

```diff
@@ -1404,22 +1411,27 @@ def resize(a, new_shape):
     """
     if isinstance(new_shape, (int, nt.integer)):
         new_shape = (new_shape,)
+
     a = ravel(a)
-    Na = len(a)
-    total_size = um.multiply.reduce(new_shape)
-    if Na == 0 or total_size == 0:
-        return mu.zeros(new_shape, a.dtype)
 
-    n_copies = int(total_size / Na)
-    extra = total_size % Na
+    new_size = 1
+    for dim_length in new_shape:
+        new_size *= dim_length
+        if dim_length < 0:
+            raise ValueError('all elements of `new_shape` must be non-negative')
 
-    if extra != 0:
-        n_copies = n_copies + 1
-        extra = Na - extra
+    if a.size == 0 or new_size == 0:
+        # First case must zero fill. The second would have repeats == 0.
+        return np.zeros_like(a, shape=new_shape)
 
-    a = concatenate((a,) * n_copies)
-    if extra > 0:
-        a = a[:-extra]
+    repeats, missing = divmod(new_size, a.size)
+
+    if missing == 0:
+        a = concatenate((a,) * repeats)
+    else:
+        # If there are some missing, repeat once more and remove the rest
+        rest = a.size - missing
+        a = concatenate((a,) * (repeats + 1))[:-rest]
```

Comment:
Some alternative spellings:

Without any tricky division:
```python
    repeats, missing = divmod(new_size, a.size)
    if missing != 0:
        repeats += 1
    a = concatenate((a,) * (repeats + 1))[:new_size]
```
Using the common formulation for round-up division:
```python
    repeats = (new_size + a.size - 1) // a.size  # ceil division
    a = concatenate((a,) * repeats)[:new_size]
```
Using the clever formulation for round-up division I wish I'd known about years ago
```python
    repeats = -(-new_size // a.size)  # ceil division
    a = concatenate((a,) * repeats)[:new_size]
```

Ok if you'd prefer to leave this as is

Reply:
Hehe, full circle ;), I first planned on doing the "common" version, then somehow decided to keep the `if: ` for the slicing and the mess happened ;). Just added your formulation the additional slicing does not matter anyway...